src/pydsl/triton.py

```diff
@@ -0,0 +1,239 @@
+from mlir.ir import InsertionPoint
+from mlir.dialects import func
+
+from pydsl.type import (
+    Bool,
+    Float,
+    Index,
+    Int,
+    Number,
+    Sign,
+    UInt32,
+)
+from pydsl.memref import MemRef
+from pydsl.protocols import lower_single
+from pydsl.compiler import ToMLIRBase
+
+
+def pydsl_type_to_triton_str(pytype: typing.Any) -> str:
+    """
+    Converts a PyDSL type to a Triton type string.
+    This is used to generate the signature for Triton functions.
+    """
+    if issubclass(pytype, MemRef):
+        return "*" + pydsl_type_to_triton_str(pytype.element_type)
+    if issubclass(pytype, Bool):
+        return "B"
+    if issubclass(pytype, Index):
+        return "i32"
+    if issubclass(pytype, Int):
+        # TODO: This currently ignores signedness, since triton-ascend does not support it.
+        sign_val = "i" if pytype.sign == Sign.UNSIGNED else "i"
+        return sign_val + str(pytype.width)
+    if issubclass(pytype, Float):
+        return "fp" + str(pytype.width)
+    # Handle constexpr (for BLOCK_SIZE, etc.)
+    if issubclass(pytype, Number):
+        return "constexpr"
+    raise TypeError(
+        f"converting {pytype.__qualname__} to a Triton type is not supported"
+    )
+
+
+def remove_uneccessary_args(func_op: func.FuncOp) -> None:
+    """
+    Removes uneccessary arguments from the funcOp.
+    This is used to clean up the MLIR after triton-adapter-opt.
+    """
+    # Find the function in the module
+    entry_block = func_op.entry_block
+    if entry_block is None:
+        raise AssertionError(
+            f"Function {func_op.sym_name.value} has no entry block."
+        )
+
+    # Collect indices of unused arguments
+    indices_to_remove = []
+    for i, arg in enumerate(entry_block.arguments):
+        # `arg.uses` is an iterable of OperationUse;
+        # if it's empty, the arg is unused. We also check if it's the first
+        # or last 6 arguments, since these are the ones added by triton-adapter-opt
+        # and we want to keep unused arguments that are user-defined.
+        if not list(arg.uses) and (
+            i == 0 or i >= len(entry_block.arguments) - 6
+        ):
+            indices_to_remove.append(i)
+
+    if not indices_to_remove:
+        return  # No unused arguments
+
+    indices_to_remove.reverse()
+
+    # Update argument attributes if present
+    if "arg_attrs" in func_op.attributes:
+        arg_attrs = func_op.arg_attrs
+        new_attrs = [
+            attr
+            for i, attr in enumerate(arg_attrs)
+            if i not in indices_to_remove
+        ]
+        func_op.arg_attrs = mlirir.ArrayAttr.get(new_attrs)
+
+    # Remove unused arguments from the entry block
+    for idx in indices_to_remove:
+        entry_block.erase_argument(idx)
+
+    # Update function type
+    old_func_type = func_op.type
+    new_inputs = [
+        ty
+        for i, ty in enumerate(old_func_type.inputs)
+        if i not in indices_to_remove
+    ]
```

Comment:
```suggestion
    new_inputs = [
        typ
        for i, typ in enumerate(old_func_type.inputs)
        if i not in indices_to_remove
    ]
```
What was wrong with your old variable name `typ`? :( (I think you used to use this before)